framework_tvb/tvb/core/services/import_service.py

```diff
@@ -314,36 +293,112 @@ def _store_imported_images(self, project):
                 if file_name.endswith(FilesHelper.TVB_FILE_EXTENSION):
                     self._populate_image(os.path.join(root, file_name), project.id)
 
-
     def import_project_operations(self, project, import_path, dt_burst_mappings=None, burst_ids_mapping=None):
         """
         This method scans provided folder and identify all operations that needs to be imported
         """
-        op_paths = self._append_tmp_to_folders_containing_operations(import_path)
-        operations = self._load_operations_from_paths(project, op_paths)
-
+        operation_paths = self._get_operation_paths(import_path)
         imported_operations = []
 
-        # Here we process each operation found
-        for operation in operations:
-            self.logger.debug("Importing operation " + str(operation))
-            old_operation_folder, _ = os.path.split(operation.import_file)
-            operation_entity, datatype_group = self.__import_operation(operation)
+        for path in operation_paths:
+            vm_paths, dt_paths = self._get_view_model_and_datatypes_paths(path)
+            if not vm_paths:
+                op_path = self._append_tmp_to_folder_containing_operation(path)
+                operation = self._load_operation_from_path(project, op_path)
+
+                self.logger.debug("Importing operation " + str(operation))
+                old_operation_folder, _ = os.path.split(operation.import_file)
+
+                operation_entity, datatype_group = self.__import_operation(operation)
+
+                # Rename operation folder with the ID of the stored operation
+                new_operation_path = FilesHelper().get_operation_folder(project.name, operation_entity.id)
+                if old_operation_folder != new_operation_path:
+                    # Delete folder of the new operation, otherwise move will fail
+                    shutil.rmtree(new_operation_path)
+                    shutil.move(old_operation_folder, new_operation_path)
+
+                operation_datatypes = self._load_datatypes_from_operation_folder(new_operation_path,
+                                                                                 operation_entity,
+                                                                                 datatype_group)
+                self._store_imported_datatypes_in_db(project, operation_datatypes, dt_burst_mappings,
+                                                     burst_ids_mapping)
+                imported_operations.append(operation_entity)
+            else:
+                view_model = self._get_main_view_model(vm_paths)
+
+                if view_model:
+                    start_date = datetime.now()
+                    alg = SimulatorAdapter().stored_adapter
```

Comment:
This should be generalized to work on any adapter type. We do not have that info in the ViewModelH5, we could add it or we could have a property on ViewModel which couples it to the proper adapter.

Reply:
I agree that this should be generalized, but the current 2 properties in each of the ViewModels seems like not the best solution
Could we build an in memory registry with Adapter - ViewModel, when we introspect at TVB startup all the adapters ? I think that is done in `tvb.config.init.initializer. _populate_algorithms method`